TypeError: Passing coroutines is forbidden, use tasks explicitly

Hi, I am trying to integrate your bluetooth integration into home assistant. Unfortunately the ESP32 bluetooth proxy was not detecting the lock (model Yale YDM-4115A with an Yale Access Bluetooth Module installed). My home assistant machine is in the rack on a VMWare ESxi host which is very hard to move close to door even for temporary testing due to router/VLAN/wiring setup. 

So I thought I will try using this to see if I am able to connect to the lock directly from my laptop first. I have updated the sample file run.py with the serial, key and key slot no obtained from rooted android phone. But I am getting an error "TypeError: Passing coroutines is forbidden, use tasks explicitly."

Could you please help me where I am going wrong?

>python --version
Python 3.11.1

>python run.py

INFO:__main__:Expected local_name: M2000LR
c:\yale\run.py:41: FutureWarning: This method will be removed in a future version of Bleak. Use the detection_callback of the BleakScanner constructor instead.
  scanner.register_detection_callback(push_lock.update_advertisement)
DEBUG:yalexs_ble.push:Waiting for advertisement callbacks for M2000LR
INFO:__main__:Started, waiting for lock to be discovered with local_name: M2000LR
DEBUG:yalexs_ble.push:M2000LR: Scheduling update to happen in 0.01 seconds
DEBUG:yalexs_ble.push:M2000LR: Update queued
DEBUG:yalexs_ble.push:M2000LR: Queued update starting
DEBUG:yalexs_ble.push:M2000LR: Starting update
DEBUG:yalexs_ble.push:M2000LR: Acquiring lock
DEBUG:yalexs_ble.push:M2000LR: Starting retry loop
DEBUG:yalexs_ble.push:M2000LR: Starting update (has_lock_info: False)
DEBUG:yalexs_ble.lock:M2000LR: Connecting to the lock
DEBUG:bleak_retry_connector:M2000LR - 98:1B:B5:01:9E:19: Connection attempt: 1
DEBUG:yalexs_ble.push:M2000LR: Scheduling update to happen in 0.01 seconds
DEBUG:yalexs_ble.push:M2000LR: Rescheduling update since one already in progress
DEBUG:yalexs_ble.push:M2000LR: Scheduling update to happen in 1.0 seconds
DEBUG:yalexs_ble.push:M2000LR: Rescheduling update since one already in progress
DEBUG:yalexs_ble.push:M2000LR: Scheduling update to happen in 1.0 seconds
DEBUG:yalexs_ble.push:M2000LR: Rescheduling update
DEBUG:yalexs_ble.push:M2000LR: Scheduling update to happen in 0.05 seconds
DEBUG:yalexs_ble.push:M2000LR: Existing update in 0.04999999998835847 seconds will happen sooner than now
DEBUG:yalexs_ble.push:M2000LR: Rescheduling update since one already in progress
DEBUG:yalexs_ble.push:M2000LR: Scheduling update to happen in 1.0 seconds
DEBUG:yalexs_ble.push:M2000LR: Rescheduling update since one already in progress
DEBUG:yalexs_ble.push:M2000LR: Scheduling update to happen in 1.0 seconds
DEBUG:yalexs_ble.push:M2000LR: Rescheduling update since one already in progress
DEBUG:yalexs_ble.push:M2000LR: Scheduling update to happen in 1.0 seconds
DEBUG:yalexs_ble.push:M2000LR: Rescheduling update since one already in progress
DEBUG:yalexs_ble.push:M2000LR: Scheduling update to happen in 1.0 seconds
DEBUG:yalexs_ble.push:M2000LR: Rescheduling update since one already in progress
DEBUG:yalexs_ble.push:M2000LR: Scheduling update to happen in 1.0 seconds
DEBUG:yalexs_ble.push:M2000LR: Rescheduling update since one already in progress
DEBUG:yalexs_ble.push:M2000LR: Scheduling update to happen in 1.0 seconds
DEBUG:yalexs_ble.push:M2000LR: Rescheduling update since one already in progress
DEBUG:yalexs_ble.push:M2000LR: Scheduling update to happen in 1.0 seconds
DEBUG:yalexs_ble.lock:M2000LR: Connected
DEBUG:yalexs_ble.session:M2000LR: Starting notify for <class 'yalexs_ble.secure_session.SecureSession'>
DEBUG:yalexs_ble.push:M2000LR: Rescheduling update since one already in progress
DEBUG:yalexs_ble.push:M2000LR: Scheduling update to happen in 1.0 seconds
DEBUG:yalexs_ble.session:M2000LR: Starting notify for <class 'yalexs_ble.session.Session'>
DEBUG:yalexs_ble.lock:M2000LR: Setting up the session
ERROR:yalexs_ble.push:M2000LR: Error updating
Traceback (most recent call last):
  File "C:\Users\xyz\AppData\Local\Programs\Python\Python311\Lib\site-packages\yalexs_ble\push.py", line 713, in _queue_update
    await self._update()
  File "C:\Users\xyz\AppData\Local\Programs\Python\Python311\Lib\site-packages\yalexs_ble\push.py", line 77, in _async_wrap_operation_lock
    return await func(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\xyz\AppData\Local\Programs\Python\Python311\Lib\site-packages\yalexs_ble\push.py", line 98, in _async_wrap_retry_bluetooth_connection_error
    return await func(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\xyz\AppData\Local\Programs\Python\Python311\Lib\site-packages\yalexs_ble\push.py", line 492, in _update
    lock = await self._ensure_connected()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\xyz\AppData\Local\Programs\Python\Python311\Lib\site-packages\yalexs_ble\push.py", line 378, in _ensure_connected
    await self._client.connect()
  File "C:\Users\xyz\AppData\Local\Programs\Python\Python311\Lib\site-packages\yalexs_ble\lock.py", line 158, in connect
    await self._setup_session()
  File "C:\Users\xyz\AppData\Local\Programs\Python\Python311\Lib\site-packages\yalexs_ble\lock.py", line 200, in _setup_session
    response = await self.secure_session.execute(self._disconnected_event, cmd)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\xyz\AppData\Local\Programs\Python\Python311\Lib\site-packages\yalexs_ble\session.py", line 216, in execute
    await asyncio.wait(
  File "C:\Users\xyz\AppData\Local\Programs\Python\Python311\Lib\asyncio\tasks.py", line 415, in wait
    raise TypeError("Passing coroutines is forbidden, use tasks explicitly.")
TypeError: Passing coroutines is forbidden, use tasks explicitly.
C:\Users\xyz\AppData\Local\Programs\Python\Python311\Lib\site-packages\yalexs_ble\push.py:744: RuntimeWarning: coroutine 'Event.wait' was never awaited
  _LOGGER.exception("%s: Error updating", self.name)
RuntimeWarning: Enable tracemalloc to get the object allocation traceback
DEBUG:yalexs_ble.session:M2000LR: Encrypted command: d70c7417889f8356f0750a7455d41fa00f01
DEBUG:yalexs_ble.session:M2000LR: Writing command to bd4ac613-0b45-11e3-8ffd-0800200c9a66 (Handle: 22): Unknown: d70c7417889f8356f0750a7455d41fa00f01
DEBUG:yalexs_ble.session:M2000LR: Waiting for response
DEBUG:yalexs_ble.session:M2000LR: Receiving response via notify: 9d86d97cf384b53df9c6dd54a3dae2280f01 (waiting=True)
DEBUG:yalexs_ble.session:M2000LR: Decrypted response via notify: 02000000cd21c4998abea606a71f955f0f01
DEBUG:yalexs_ble.secure_session:M2000LR: Response security checksum: 1603608487, expected: 1603608487
DEBUG:yalexs_ble.session:M2000LR: Got response: 02000000cd21c4998abea606a71f955f0f01
DEBUG:yalexs_ble.push:M2000LR: Update queued
DEBUG:yalexs_ble.push:M2000LR: Queued update starting
DEBUG:yalexs_ble.push:M2000LR: Starting update
DEBUG:yalexs_ble.push:M2000LR: Acquiring lock
DEBUG:yalexs_ble.push:M2000LR: Starting retry loop
DEBUG:yalexs_ble.push:M2000LR: Starting update (has_lock_info: False)
DEBUG:yalexs_ble.lock:M2000LR: Probing the lock
DEBUG:yalexs_ble.lock:M2000LR: Executing status
ERROR:yalexs_ble.push:M2000LR: Error updating
Traceback (most recent call last):
  File "C:\Users\xyz\AppData\Local\Programs\Python\Python311\Lib\site-packages\yalexs_ble\push.py", line 713, in _queue_update
    await self._update()
  File "C:\Users\xyz\AppData\Local\Programs\Python\Python311\Lib\site-packages\yalexs_ble\push.py", line 77, in _async_wrap_operation_lock
    return await func(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\xyz\AppData\Local\Programs\Python\Python311\Lib\site-packages\yalexs_ble\push.py", line 98, in _async_wrap_retry_bluetooth_connection_error
    return await func(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\xyz\AppData\Local\Programs\Python\Python311\Lib\site-packages\yalexs_ble\push.py", line 495, in _update
    state = await lock.status()
            ^^^^^^^^^^^^^^^^^^^
  File "C:\Users\xyz\AppData\Local\Programs\Python\Python311\Lib\site-packages\yalexs_ble\lock.py", line 298, in status
    response = await self._execute_command(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\xyz\AppData\Local\Programs\Python\Python311\Lib\site-packages\yalexs_ble\lock.py", line 271, in _execute_command
    response = await self.session.execute(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\xyz\AppData\Local\Programs\Python\Python311\Lib\site-packages\yalexs_ble\session.py", line 216, in execute
    await asyncio.wait(
  File "C:\Users\xyz\AppData\Local\Programs\Python\Python311\Lib\asyncio\tasks.py", line 415, in wait
    raise TypeError("Passing coroutines is forbidden, use tasks explicitly.")
TypeError: Passing coroutines is forbidden, use tasks explicitly.
ERROR:asyncio:Task exception was never retrieved
future: <Task finished name='Task-25' coro=<Session._write() done, defined at C:\Users\xyz\AppData\Local\Programs\Python\Python311\Lib\site-packages\yalexs_ble\session.py:102> exception=AssertionError('Cipher not set')>
Traceback (most recent call last):
  File "C:\Users\xyz\AppData\Local\Programs\Python\Python311\Lib\site-packages\yalexs_ble\session.py", line 105, in _write
    return await self._locked_write(command)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\xyz\AppData\Local\Programs\Python\Python311\Lib\site-packages\yalexs_ble\session.py", line 139, in _locked_write
    assert self.cipher_encrypt is not None, "Cipher not set"  # nosec
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: Cipher not set
DEBUG:yalexs_ble.lock:M2000LR: Disconnected from lock callback
Traceback (most recent call last):
  File "C:\Users\xyz\AppData\Local\Programs\Python\Python311\Lib\asyncio\runners.py", line 118, in run
Exception ignored in: <function BleakScannerWinRT.start.<locals>.<lambda> at 0x0000019F1208BCE0>
Traceback (most recent call last):
  File "C:\Users\xyz\AppData\Local\Programs\Python\Python311\Lib\site-packages\bleak\backends\winrt\scanner.py", line 245, in <lambda>
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\xyz\AppData\Local\Programs\Python\Python311\Lib\asyncio\base_events.py", line 653, in run_until_complete
    lambda s, e: event_loop.call_soon_threadsafe(self._received_handler, s, e)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\xyz\AppData\Local\Programs\Python\Python311\Lib\asyncio\base_events.py", line 806, in call_soon_threadsafe
    return future.result()
           ^^^^^^^^^^^^^^^
    self._check_closed()
  File "c:\Data\ESP8266\yale\run.py", line 48, in run
  File "C:\Users\xyz\AppData\Local\Programs\Python\Python311\Lib\asyncio\base_events.py", line 519, in _check_closed
    await asyncio.sleep(1000000)
    raise RuntimeError('Event loop is closed')
RuntimeError: Event loop is closed
  File "C:\Users\xyz\AppData\Local\Programs\Python\Python311\Lib\asyncio\tasks.py", line 639, in sleep
    return await future
           ^^^^^^^^^^^^
asyncio.exceptions.CancelledError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "c:\Data\ESP8266\yale\run.py", line 57, in <module>
    asyncio.run(run())
  File "C:\Users\xyz\AppData\Local\Programs\Python\Python311\Lib\asyncio\runners.py", line 190, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "C:\Users\xyz\AppData\Local\Programs\Python\Python311\Lib\asyncio\runners.py", line 123, in run
    raise KeyboardInterrupt()

--- a/src/yalexs_ble/session.py
+++ b/src/yalexs_ble/session.py
@@ -215,10 +215,13 @@ class Session:
         assert self.cipher_encrypt is not None, "Cipher not set"  # nosec
         self._write_checksum(command)
         write_task = asyncio.create_task(self._write(command))
+        disconnect_task = asyncio.create_task(self._disconnected_event.wait())
         await asyncio.wait(
-            [write_task, self._disconnected_event.wait()],
+            (write_task, disconnect_task),
             return_when=asyncio.FIRST_COMPLETED,
         )
+        if disconnect_task.done():
+            write_task.cancel()
         if write_task.done():
             try:
                 return await write_task

